mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,1046 @@
+#include <iostream>
+#include <set>
+#include <unordered_map>
+
+#define cache_line_size 64
+
+using namespace mlir;
+
+namespace {
+struct LoopInterchange : public AffineLoopInterchangeBase<LoopInterchange> {
+  void runOnFunction() override;
+  void handleImperfectlyNestedAffineLoops(Operation &funcOp);
+
+protected:
+  void getAffineAccessMatrices(
+      AffineForOp rootForOp,
+      llvm::SmallVector<AffineForOp, 0> &AffineForOpLoopNest,
+      llvm::DenseMap<Value, int64_t> &loopVarIdMap);
+  void getDependencesPresentInAffineLoopNest(
+      AffineForOp forOp, unsigned maxLoopDepth,
+      std::vector<llvm::SmallVector<int64_t, 8>> *depCompsMatrix);
+  void getAffineLoopCarriedDependencies(
+      std::vector<llvm::SmallVector<int64_t, 8>> *dependenceMatrix);
+  void buildReferenceGroups(AffineForOp innermostForOp, unsigned maxDepth,
+                            unsigned forOpLevelInOriginalLoopNest);
+  void getCacheLineAccessCounts(
+      llvm::SmallVector<AffineForOp, 0> &AffineForOpLoopNest);
+  long double getTemporalReuseCost(llvm::SmallVector<unsigned, 0> &permutation);
+  llvm::SmallVector<unsigned, 0> getBestPermutation(
+      llvm::SmallVector<llvm::SmallVector<unsigned, 0>, 0> &validPermutations,
+      llvm::SmallVector<AffineForOp, 0> &AffineForOpLoopNest);
+  void clear(llvm::SmallVector<AffineForOp, 0> &AffineForOpLoopNest,
+             llvm::DenseMap<Value, int64_t> &loopVarIdMap, int *loopVarIdCount);
+
+  llvm::DenseMap<Operation *, llvm::SmallVector<std::vector<int64_t>, 0>>
+      loopAccessMatrices; // For each memory access Ax+B in the loopnest, it is
+                          // the collection of A's indexed by respective
+                          // load/store operation
+  llvm::DenseMap<Operation *, std::vector<int64_t>>
+      constVector; // For a memory access Ax+B, this represents B
+  llvm::DenseMap<Operation *, unsigned>
+      dataSize; // The size of data in memref access
+  llvm::SmallVector<unsigned, 0>
+      loopIterationCounts; // Iteration counts for each AffineForOp
+  std::vector<int>
+      loopCarriedDependenceVector; // Loop Carried Dependency vector. A value of
+                                   // "1" at index i represents there is a
+                                   // dependency carried on the loop at depth i
+                                   // in the loopnest
+  llvm::DenseMap<AffineForOp *, long double>
+      numberOfCacheLinesAccessed; // Number of cache lines accessed by each
+                                  // AffineForOp if they were the innermost and
+                                  // only loop in the loopnest
+  std::vector<std::set<Operation *>> referenceGroups;
+};
+} // namespace
+
+/// <summary>
+/// Performs a walk over the loopnest rooted at "forOp" to check if there are
+/// any intermediate "Affine-if" statements in the loopnest.
+/// </summary>
+/// <param name="forOp"></param>
+/// <returns>Returns True if an "Affine-if" statement found.</returns>
```

Comment:
Just /// comments is enough. The tags aren't needed - please see other code for reference. Include parameter names in backticks.